include/wlr/types/wlr_text_input.h

```diff
@@ -0,0 +1,77 @@
+#ifndef WLR_TYPES_WLR_TEXT_INPUT_H
+#define WLR_TYPES_WLR_TEXT_INPUT_H
+
+#include <wayland-server.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_surface.h>
+
+struct wlr_text_input_properties {
+	struct {
+		char *text; // NULL is allowed and equivalent to empty string
+		uint32_t cursor;
+		uint32_t anchor;
+	} surrounding;
+
+	uint32_t text_change_cause;
+
+	struct {
+		uint32_t hint;
+		uint32_t purpose;
+	} content_type;
+
+	struct {
+		int32_t x;
+		int32_t y;
+		int32_t width;
+		int32_t height;
+	} cursor_rectangle;
+};
+
+struct wlr_text_input {
+	struct wlr_seat *seat; // used only for unlinking seat->text_input; perhaps better to use the destroy signal
+	struct wl_resource *resource;
+	struct wlr_surface *focused_surface;
+	struct wlr_text_input_properties pending;
+	struct wlr_text_input_properties current;
+	uint32_t current_serial; // next to send
+	bool pending_enabled;
+	bool current_enabled;
+
+	struct wl_list link;
+
+	struct wl_listener seat_destroy_listener;
+
+	struct {
+		struct wl_signal enable; // (struct wlr_text_input*)
+		struct wl_signal commit; // (struct wlr_text_input*)
+		struct wl_signal disable; // (struct wlr_text_input*)
+		struct wl_signal destroy; // (struct wlr_text_input*)
+	} events;
+};
+
+struct wlr_text_input_manager {
+	struct wl_global *wl_global;
+	struct wl_list text_inputs; // wlr_text_input::resource::link
+
+	struct {
+		struct wl_signal text_input; // (struct wlr_text_input*)
+	} events;
```

Comment:
Should have a `destroy` event

Reply:
I mean a destroy event on the manager. The manager gets destroyed with the display. If the compositor has objects that depend on it it needs to destroy them with the manager (it can't destroy them with the display because the destroy listener of the manager may be triggered before the destroy listener of the objects, this has caused issues in the past).